the unit tests have failed:
```
 	 # github.com/cilium/cilium/pkg/fqdn/dnsproxy [github.com/cilium/cilium/pkg/fqdn/dnsproxy.test]
	 pkg/fqdn/dnsproxy/proxy_test.go:184:36: not enough arguments in call to testSelectorCache.UpdateIdentities
	 	have ("github.com/cilium/cilium/pkg/identity/cache".IdentityCache, nil)
	 	want ("github.com/cilium/cilium/pkg/identity/cache".IdentityCache, "github.com/cilium/cilium/pkg/identity/cache".IdentityCache, *sync.WaitGroup)
	 make: *** [tests-privileged] Error 1
```

--- a/daemon/cmd/fqdn.go
+++ b/daemon/cmd/fqdn.go
@@ -97,13 +97,14 @@ func identitiesForFQDNSelectorIPs(selectorsWithIPsToUpdate map[policyApi.FQDNSel
 	return selectorIdentitySliceMapping, newlyAllocatedIdentities, nil
 }
 
-func (d *Daemon) updateSelectorCacheFQDNs(ctx context.Context, selectors map[policyApi.FQDNSelector][]*identity.Identity, selectorsWithoutIPs []policyApi.FQDNSelector) (wg *sync.WaitGroup) {
+func (d *Daemon) updateSelectorCacheFQDNs(ctx context.Context, selectors map[policyApi.FQDNSelector][]*identity.Identity, selectorsWithoutIPs []policyApi.FQDNSelector) *sync.WaitGroup {
 	// There may be nothing to update - in this case, we exit and do not need
 	// to trigger policy updates for all endpoints.
 	if len(selectors) == 0 && len(selectorsWithoutIPs) == 0 {
 		return &sync.WaitGroup{}
 	}
 
+	notifyWg := &sync.WaitGroup{}
 	// Update mapping of selector to set of IPs in selector cache.
 	for selector, identitySlice := range selectors {
 		log.WithFields(logrus.Fields{
@@ -114,7 +115,7 @@ func (d *Daemon) updateSelectorCacheFQDNs(ctx context.Context, selectors map[pol
 			// Nil check here? Hopefully not necessary...
 			numIds = append(numIds, numId.ID)
 		}
-		d.policy.GetSelectorCache().UpdateFQDNSelector(selector, numIds)
+		d.policy.GetSelectorCache().UpdateFQDNSelector(selector, numIds, notifyWg)
 	}
 
 	if len(selectorsWithoutIPs) > 0 {
@@ -125,10 +126,10 @@ func (d *Daemon) updateSelectorCacheFQDNs(ctx context.Context, selectors map[pol
 		log.WithFields(logrus.Fields{
 			"fqdnSelectors": selectorsWithoutIPs,
 		}).Debug("removing all identities from FQDN selectors")
-		d.policy.GetSelectorCache().RemoveIdentitiesFQDNSelectors(selectorsWithoutIPs)
+		d.policy.GetSelectorCache().RemoveIdentitiesFQDNSelectors(selectorsWithoutIPs, notifyWg)
 	}
 
-	return d.endpointManager.UpdatePolicyMaps(ctx)
+	return d.endpointManager.UpdatePolicyMaps(ctx, notifyWg)
 }
 
 // bootstrapFQDN initializes the toFQDNs related subsystems: dnsNameManager and the DNS proxy.

--- a/daemon/cmd/policy.go
+++ b/daemon/cmd/policy.go
@@ -95,7 +95,10 @@ func (d *Daemon) TriggerPolicyUpdates(force bool, reason string) {
 // The caller is responsible for making sure the same identity is not
 // present in both 'added' and 'deleted'.
 func (d *Daemon) UpdateIdentities(added, deleted cache.IdentityCache) {
-	d.policy.GetSelectorCache().UpdateIdentities(added, deleted)
+	wg := &sync.WaitGroup{}
+	d.policy.GetSelectorCache().UpdateIdentities(added, deleted, wg)
+	// Wait for update propagation to endpoints before triggering policy updates
+	wg.Wait()
 	d.TriggerPolicyUpdates(false, "one or more identities created or deleted")
 }
 

--- a/pkg/endpointmanager/manager.go
+++ b/pkg/endpointmanager/manager.go
@@ -145,7 +145,7 @@ func waitForProxyCompletions(proxyWaitGroup *completion.WaitGroup) error {
 
 // UpdatePolicyMaps returns a WaitGroup which is signaled upon once all endpoints
 // have had their PolicyMaps updated against the Endpoint's desired policy state.
-func (mgr *EndpointManager) UpdatePolicyMaps(ctx context.Context) *sync.WaitGroup {
+func (mgr *EndpointManager) UpdatePolicyMaps(ctx context.Context, notifyWg *sync.WaitGroup) *sync.WaitGroup {
 	var epWG sync.WaitGroup
 	var wg sync.WaitGroup
 
@@ -169,6 +169,8 @@ func (mgr *EndpointManager) UpdatePolicyMaps(ctx context.Context) *sync.WaitGrou
 	// TODO: bound by number of CPUs?
 	for _, ep := range eps {
 		go func(ep *endpoint.Endpoint) {
+			// Proceed only after all notifications have been delivered to endpoints
+			notifyWg.Wait()
 			if err := ep.ApplyPolicyMapChanges(proxyWaitGroup); err != nil {
 				ep.Logger("endpointmanager").WithError(err).Warning("Failed to apply policy map changes. These will be re-applied in future updates.")
 			}

--- a/pkg/fqdn/dnsproxy/proxy_test.go
+++ b/pkg/fqdn/dnsproxy/proxy_test.go
@@ -22,6 +22,7 @@ import (
 	"encoding/json"
 	"fmt"
 	"net"
+	"sync"
 	"testing"
 	"time"
 
@@ -181,12 +182,14 @@ var (
 
 func (s *DNSProxyTestSuite) SetUpTest(c *C) {
 	// Add these identities
+	wg := &sync.WaitGroup{}
 	testSelectorCache.UpdateIdentities(cache.IdentityCache{
 		dstID1: labels.Labels{"Dst1": labels.NewLabel("Dst1", "test", labels.LabelSourceK8s)}.LabelArray(),
 		dstID2: labels.Labels{"Dst2": labels.NewLabel("Dst2", "test", labels.LabelSourceK8s)}.LabelArray(),
 		dstID3: labels.Labels{"Dst3": labels.NewLabel("Dst3", "test", labels.LabelSourceK8s)}.LabelArray(),
 		dstID4: labels.Labels{"Dst4": labels.NewLabel("Dst4", "test", labels.LabelSourceK8s)}.LabelArray(),
-	}, nil)
+	}, nil, wg)
+	wg.Wait()
 
 	s.repo = policy.NewPolicyRepository(nil, nil)
 	s.dnsTCPClient = &dns.Client{Net: "tcp", Timeout: time.Second, SingleInflight: true}

--- a/pkg/policy/l4.go
+++ b/pkg/policy/l4.go
@@ -34,7 +34,7 @@ import (
 	"github.com/cilium/cilium/pkg/policy/api"
 	"github.com/cilium/cilium/pkg/policy/trafficdirection"
 	"github.com/cilium/cilium/pkg/u8proto"
-	"github.com/cilium/proxy/go/cilium/api"
+	cilium "github.com/cilium/proxy/go/cilium/api"
 
 	"github.com/sirupsen/logrus"
 )
@@ -426,8 +426,8 @@ func (l4Filter *L4Filter) ToMapState(policyOwner PolicyOwner, direction trafficd
 }
 
 // IdentitySelectionUpdated implements CachedSelectionUser interface
-// This call is made while holding name manager and selector cache
-// locks, must beware of deadlocking!
+// This call is made from a single goroutine in FIFO order to keep add
+// and delete events ordered properly. No locks are held.
 //
 // The caller is responsible for making sure the same identity is not
 // present in both 'added' and 'deleted'.

--- a/pkg/policy/resolve_deny_test.go
+++ b/pkg/policy/resolve_deny_test.go
@@ -17,7 +17,7 @@
 package policy
 
 import (
-	"time"
+	"sync"
 
 	"github.com/cilium/cilium/pkg/checker"
 	"github.com/cilium/cilium/pkg/identity"
@@ -357,8 +357,9 @@ func (ds *PolicyTestSuite) TestMapStateWithIngressDenyWildcard(c *C) {
 	added1 := cache.IdentityCache{
 		identity.NumericIdentity(192): labels.ParseSelectLabelArray("id=resolve_test_1"),
 	}
-	testSelectorCache.UpdateIdentities(added1, nil)
-	time.Sleep(100 * time.Millisecond)
+	wg := &sync.WaitGroup{}
+	testSelectorCache.UpdateIdentities(added1, nil, wg)
+	wg.Wait()
 	c.Assert(policy.policyMapChanges.changes, HasLen, 0)
 
 	// Have to remove circular reference before testing to avoid an infinite loop
@@ -433,17 +434,19 @@ func (ds *PolicyTestSuite) TestMapStateWithIngressDeny(c *C) {
 		identity.NumericIdentity(193): labels.ParseSelectLabelArray("id=resolve_test_1", "num=2"),
 		identity.NumericIdentity(194): labels.ParseSelectLabelArray("id=resolve_test_1", "num=3"),
 	}
-	testSelectorCache.UpdateIdentities(added1, nil)
+	wg := &sync.WaitGroup{}
+	testSelectorCache.UpdateIdentities(added1, nil, wg)
 	// Cleanup the identities from the testSelectorCache
-	defer testSelectorCache.UpdateIdentities(nil, added1)
-	time.Sleep(100 * time.Millisecond)
+	defer testSelectorCache.UpdateIdentities(nil, added1, wg)
+	wg.Wait()
 	c.Assert(policy.policyMapChanges.changes, HasLen, 3)
 
 	deleted1 := cache.IdentityCache{
 		identity.NumericIdentity(193): labels.ParseSelectLabelArray("id=resolve_test_1", "num=2"),
 	}
-	testSelectorCache.UpdateIdentities(nil, deleted1)
-	time.Sleep(100 * time.Millisecond)
+	wg = &sync.WaitGroup{}
+	testSelectorCache.UpdateIdentities(nil, deleted1, wg)
+	wg.Wait()
 	c.Assert(policy.policyMapChanges.changes, HasLen, 4)
 
 	cachedSelectorWorld := testSelectorCache.FindCachedIdentitySelector(api.ReservedEndpointSelectors[labels.IDNameWorld])

--- a/pkg/policy/resolve_deny_test.go
+++ b/pkg/policy/resolve_deny_test.go
@@ -17,6 +17,8 @@
 package policy
 
 import (
+	"time"
+
 	"github.com/cilium/cilium/pkg/checker"
 	"github.com/cilium/cilium/pkg/identity"
 	"github.com/cilium/cilium/pkg/identity/cache"
@@ -356,6 +358,7 @@ func (ds *PolicyTestSuite) TestMapStateWithIngressDenyWildcard(c *C) {
 		identity.NumericIdentity(192): labels.ParseSelectLabelArray("id=resolve_test_1"),
 	}
 	testSelectorCache.UpdateIdentities(added1, nil)
+	time.Sleep(100 * time.Millisecond)
 	c.Assert(policy.policyMapChanges.changes, HasLen, 0)
 
 	// Have to remove circular reference before testing to avoid an infinite loop
@@ -433,12 +436,14 @@ func (ds *PolicyTestSuite) TestMapStateWithIngressDeny(c *C) {
 	testSelectorCache.UpdateIdentities(added1, nil)
 	// Cleanup the identities from the testSelectorCache
 	defer testSelectorCache.UpdateIdentities(nil, added1)
+	time.Sleep(100 * time.Millisecond)
 	c.Assert(policy.policyMapChanges.changes, HasLen, 3)
 
 	deleted1 := cache.IdentityCache{
 		identity.NumericIdentity(193): labels.ParseSelectLabelArray("id=resolve_test_1", "num=2"),
 	}
 	testSelectorCache.UpdateIdentities(nil, deleted1)
+	time.Sleep(100 * time.Millisecond)
 	c.Assert(policy.policyMapChanges.changes, HasLen, 4)
 
 	cachedSelectorWorld := testSelectorCache.FindCachedIdentitySelector(api.ReservedEndpointSelectors[labels.IDNameWorld])

--- a/pkg/policy/resolve_test.go
+++ b/pkg/policy/resolve_test.go
@@ -20,7 +20,6 @@ import (
 	"fmt"
 	"sync"
 	"testing"
-	"time"
 
 	"github.com/cilium/cilium/pkg/checker"
 	"github.com/cilium/cilium/pkg/identity"
@@ -202,10 +201,11 @@ func bootstrapRepo(ruleGenFunc func(int) api.Rules, numRules int, c *C) *Reposit
 	mgr := cache.NewCachingIdentityAllocator(&allocator.IdentityAllocatorOwnerMock{})
 	testRepo := NewPolicyRepository(mgr.GetIdentityCache(), nil)
 
-	var wg sync.WaitGroup
 	SetPolicyEnabled(option.DefaultEnforcement)
 	GenerateNumIdentities(3000)
-	testSelectorCache.UpdateIdentities(identityCache, nil)
+	wg := &sync.WaitGroup{}
+	testSelectorCache.UpdateIdentities(identityCache, nil, wg)
+	wg.Wait()
 	testRepo.selectorCache = testSelectorCache
 	rulez, _ := testRepo.AddList(ruleGenFunc(numRules))
 
@@ -217,7 +217,8 @@ func bootstrapRepo(ruleGenFunc func(int) api.Rules, numRules int, c *C) *Reposit
 	})
 
 	epsToRegen := NewEndpointSet(nil)
-	rulez.UpdateRulesEndpointsCaches(epSet, epsToRegen, &wg)
+	wg = &sync.WaitGroup{}
+	rulez.UpdateRulesEndpointsCaches(epSet, epsToRegen, wg)
 	wg.Wait()
 
 	c.Assert(epSet.Len(), Equals, 0)
@@ -519,8 +520,9 @@ func (ds *PolicyTestSuite) TestMapStateWithIngressWildcard(c *C) {
 	added1 := cache.IdentityCache{
 		identity.NumericIdentity(192): labels.ParseSelectLabelArray("id=resolve_test_1"),
 	}
-	testSelectorCache.UpdateIdentities(added1, nil)
-	time.Sleep(100 * time.Millisecond)
+	wg := &sync.WaitGroup{}
+	testSelectorCache.UpdateIdentities(added1, nil, wg)
+	wg.Wait()
 	c.Assert(policy.policyMapChanges.changes, HasLen, 0)
 
 	// Have to remove circular reference before testing to avoid an infinite loop
@@ -597,17 +599,19 @@ func (ds *PolicyTestSuite) TestMapStateWithIngress(c *C) {
 		identity.NumericIdentity(193): labels.ParseSelectLabelArray("id=resolve_test_1", "num=2"),
 		identity.NumericIdentity(194): labels.ParseSelectLabelArray("id=resolve_test_1", "num=3"),
 	}
-	testSelectorCache.UpdateIdentities(added1, nil)
+	wg := &sync.WaitGroup{}
+	testSelectorCache.UpdateIdentities(added1, nil, wg)
 	// Cleanup the identities from the testSelectorCache
-	defer testSelectorCache.UpdateIdentities(nil, added1)
-	time.Sleep(100 * time.Millisecond)
+	defer testSelectorCache.UpdateIdentities(nil, added1, wg)
+	wg.Wait()
 	c.Assert(policy.policyMapChanges.changes, HasLen, 3)
 
 	deleted1 := cache.IdentityCache{
 		identity.NumericIdentity(193): labels.ParseSelectLabelArray("id=resolve_test_1", "num=2"),
 	}
-	testSelectorCache.UpdateIdentities(nil, deleted1)
-	time.Sleep(100 * time.Millisecond)
+	wg = &sync.WaitGroup{}
+	testSelectorCache.UpdateIdentities(nil, deleted1, wg)
+	wg.Wait()
 	c.Assert(policy.policyMapChanges.changes, HasLen, 4)
 
 	cachedSelectorWorld := testSelectorCache.FindCachedIdentitySelector(api.ReservedEndpointSelectors[labels.IDNameWorld])

--- a/pkg/policy/resolve_test.go
+++ b/pkg/policy/resolve_test.go
@@ -20,6 +20,7 @@ import (
 	"fmt"
 	"sync"
 	"testing"
+	"time"
 
 	"github.com/cilium/cilium/pkg/checker"
 	"github.com/cilium/cilium/pkg/identity"
@@ -519,6 +520,7 @@ func (ds *PolicyTestSuite) TestMapStateWithIngressWildcard(c *C) {
 		identity.NumericIdentity(192): labels.ParseSelectLabelArray("id=resolve_test_1"),
 	}
 	testSelectorCache.UpdateIdentities(added1, nil)
+	time.Sleep(100 * time.Millisecond)
 	c.Assert(policy.policyMapChanges.changes, HasLen, 0)
 
 	// Have to remove circular reference before testing to avoid an infinite loop
@@ -598,12 +600,14 @@ func (ds *PolicyTestSuite) TestMapStateWithIngress(c *C) {
 	testSelectorCache.UpdateIdentities(added1, nil)
 	// Cleanup the identities from the testSelectorCache
 	defer testSelectorCache.UpdateIdentities(nil, added1)
+	time.Sleep(100 * time.Millisecond)
 	c.Assert(policy.policyMapChanges.changes, HasLen, 3)
 
 	deleted1 := cache.IdentityCache{
 		identity.NumericIdentity(193): labels.ParseSelectLabelArray("id=resolve_test_1", "num=2"),
 	}
 	testSelectorCache.UpdateIdentities(nil, deleted1)
+	time.Sleep(100 * time.Millisecond)
 	c.Assert(policy.policyMapChanges.changes, HasLen, 4)
 
 	cachedSelectorWorld := testSelectorCache.FindCachedIdentitySelector(api.ReservedEndpointSelectors[labels.IDNameWorld])

--- a/pkg/policy/selectorcache.go
+++ b/pkg/policy/selectorcache.go
@@ -19,6 +19,7 @@ import (
 	"encoding/json"
 	"sort"
 	"strings"
+	"sync"
 	"sync/atomic"
 	"unsafe"
 
@@ -152,7 +153,7 @@ type identitySelector interface {
 	removeUser(CachedSelectionUser, identityNotifier) (last bool)
 
 	// This may be called while the NameManager lock is held
-	notifyUsers(added, deleted []identity.NumericIdentity)
+	notifyUsers(sc *SelectorCache, added, deleted []identity.NumericIdentity)
 
 	numUsers() int
 }
@@ -183,6 +184,17 @@ func getIdentityCache(ids cache.IdentityCache) scIdentityCache {
 	return idCache
 }
 
+// userNotification stores the information needed to call
+// IdentitySelectionUpdated callbacks to notify users of selector's
+// identity changes. These are queued to be able to call the callbacks
+// in FIFO order while not holding any locks.
+type userNotification struct {
+	user     CachedSelectionUser
+	selector CachedSelector
+	added    []identity.NumericIdentity
+	deleted  []identity.NumericIdentity
+}
+
 // SelectorCache caches identities, identity selectors, and the
 // subsets of identities each selector selects.
 type SelectorCache struct {
@@ -197,6 +209,14 @@ type SelectorCache struct {
 	selectors map[string]identitySelector
 
 	localIdentityNotifier identityNotifier
+
+	// userCond is a condition variable for receiving signals
+	// about addition of new elements in userNotes
+	userCond *sync.Cond
+	// userMutex protects userNotes and is linked to userCond
+	userMutex lock.Mutex
+	// userNotes holds a FIFO list of user notifications to be made
+	userNotes []userNotification
 }
 
 // GetModel returns the API model of the SelectorCache.
@@ -223,12 +243,46 @@ func (sc *SelectorCache) GetModel() models.SelectorCache {
 	return selCacheMdl
 }
 
+func (sc *SelectorCache) handleUserNotifications() {
+	for {
+		sc.userMutex.Lock()
+		for len(sc.userNotes) == 0 {
+			sc.userCond.Wait()
+		}
+		// get the current batch of notifications and release the lock so that SelectorCache
+		// can't block on userMutex while we call IdentitySelectionUpdated callbacks below.
+		notifications := sc.userNotes
+		sc.userNotes = nil
+		sc.userMutex.Unlock()
+
+		for _, n := range notifications {
+			n.user.IdentitySelectionUpdated(n.selector, n.added, n.deleted)
+		}
+	}
+}
+
+func (sc *SelectorCache) queueUserNotification(user CachedSelectionUser, selector CachedSelector, added, deleted []identity.NumericIdentity) {
+	sc.userMutex.Lock()
+	sc.userNotes = append(sc.userNotes, userNotification{
+		user:     user,
+		selector: selector,
+		added:    added,
+		deleted:  deleted,
+	})
+	sc.userMutex.Unlock()
+	sc.userCond.Signal()
+}
+
 // NewSelectorCache creates a new SelectorCache with the given identities.
 func NewSelectorCache(ids cache.IdentityCache) *SelectorCache {
-	return &SelectorCache{
+	sc := &SelectorCache{
 		idCache:   getIdentityCache(ids),
 		selectors: make(map[string]identitySelector),
 	}
+	sc.userCond = sync.NewCond(&sc.userMutex)
+	go sc.handleUserNotifications()
+
+	return sc
 }
 
 // SetLocalIdentityNotifier injects the provided identityNotifier into the
@@ -375,10 +429,10 @@ type fqdnSelector struct {
 //
 // The caller is responsible for making sure the same identity is not
 // present in both 'added' and 'deleted'.
-func (f *fqdnSelector) notifyUsers(added, deleted []identity.NumericIdentity) {
+func (f *fqdnSelector) notifyUsers(sc *SelectorCache, added, deleted []identity.NumericIdentity) {
 	for user := range f.users {
 		// pass 'f' to the user as '*fqdnSelector'
-		user.IdentitySelectionUpdated(f, added, deleted)
+		sc.queueUserNotification(user, f, added, deleted)
 	}
 }
 
@@ -434,10 +488,10 @@ type labelIdentitySelector struct {
 //
 // The caller is responsible for making sure the same identity is not
 // present in both 'added' and 'deleted'.
-func (l *labelIdentitySelector) notifyUsers(added, deleted []identity.NumericIdentity) {
+func (l *labelIdentitySelector) notifyUsers(sc *SelectorCache, added, deleted []identity.NumericIdentity) {
 	for user := range l.users {
 		// pass 'l' to the user as '*labelIdentitySelector'
-		user.IdentitySelectionUpdated(l, added, deleted)
+		sc.queueUserNotification(user, l, added, deleted)
 	}
 }
 
@@ -567,7 +621,7 @@ func (sc *SelectorCache) updateFQDNSelector(fqdnSelec api.FQDNSelector, identiti
 	// getting the CIDR identities which correspond to this FQDNSelector. This
 	// is the primary difference here between FQDNSelector and IdentitySelector.
 	fqdnSel.updateSelections()
-	fqdnSel.notifyUsers(added, deleted) // disjoint sets, see the comment above
+	fqdnSel.notifyUsers(sc, added, deleted) // disjoint sets, see the comment above
 }
 
 // AddFQDNSelector adds the given api.FQDNSelector in to the selector cache. If
@@ -826,7 +880,7 @@ func (sc *SelectorCache) UpdateIdentities(added, deleted cache.IdentityCache) {
 				}
 				if len(dels)+len(adds) > 0 {
 					idSel.updateSelections()
-					idSel.notifyUsers(adds, dels)
+					idSel.notifyUsers(sc, adds, dels)
 				}
 			case *fqdnSelector:
 				// This is a no-op right now. We don't encode in the identities

--- a/pkg/policy/selectorcache.go
+++ b/pkg/policy/selectorcache.go
@@ -152,8 +152,11 @@ type identitySelector interface {
 	// Called with NameManager and SelectorCache locks held
 	removeUser(CachedSelectionUser, identityNotifier) (last bool)
 
-	// This may be called while the NameManager lock is held
-	notifyUsers(sc *SelectorCache, added, deleted []identity.NumericIdentity)
+	// This may be called while the NameManager lock is held. wg.Wait()
+	// returns after user notifications have been completed, which may require
+	// taking Endpoint and SelectorCache locks, so these locks must not be
+	// held when calling wg.Wait().
+	notifyUsers(sc *SelectorCache, added, deleted []identity.NumericIdentity, wg *sync.WaitGroup)
 
 	numUsers() int
 }
@@ -193,6 +196,7 @@ type userNotification struct {
 	selector CachedSelector
 	added    []identity.NumericIdentity
 	deleted  []identity.NumericIdentity
+	wg       *sync.WaitGroup
 }
 
 // SelectorCache caches identities, identity selectors, and the
@@ -257,17 +261,20 @@ func (sc *SelectorCache) handleUserNotifications() {
 
 		for _, n := range notifications {
 			n.user.IdentitySelectionUpdated(n.selector, n.added, n.deleted)
+			n.wg.Done()
 		}
 	}
 }
 
-func (sc *SelectorCache) queueUserNotification(user CachedSelectionUser, selector CachedSelector, added, deleted []identity.NumericIdentity) {
+func (sc *SelectorCache) queueUserNotification(user CachedSelectionUser, selector CachedSelector, added, deleted []identity.NumericIdentity, wg *sync.WaitGroup) {
+	wg.Add(1)
 	sc.userMutex.Lock()
 	sc.userNotes = append(sc.userNotes, userNotification{
 		user:     user,
 		selector: selector,
 		added:    added,
 		deleted:  deleted,
+		wg:       wg,
 	})
 	sc.userMutex.Unlock()
 	sc.userCond.Signal()
@@ -429,10 +436,10 @@ type fqdnSelector struct {
 //
 // The caller is responsible for making sure the same identity is not
 // present in both 'added' and 'deleted'.
-func (f *fqdnSelector) notifyUsers(sc *SelectorCache, added, deleted []identity.NumericIdentity) {
+func (f *fqdnSelector) notifyUsers(sc *SelectorCache, added, deleted []identity.NumericIdentity, wg *sync.WaitGroup) {
 	for user := range f.users {
 		// pass 'f' to the user as '*fqdnSelector'
-		sc.queueUserNotification(user, f, added, deleted)
+		sc.queueUserNotification(user, f, added, deleted, wg)
 	}
 }
 
@@ -488,10 +495,10 @@ type labelIdentitySelector struct {
 //
 // The caller is responsible for making sure the same identity is not
 // present in both 'added' and 'deleted'.
-func (l *labelIdentitySelector) notifyUsers(sc *SelectorCache, added, deleted []identity.NumericIdentity) {
+func (l *labelIdentitySelector) notifyUsers(sc *SelectorCache, added, deleted []identity.NumericIdentity, wg *sync.WaitGroup) {
 	for user := range l.users {
 		// pass 'l' to the user as '*labelIdentitySelector'
-		sc.queueUserNotification(user, l, added, deleted)
+		sc.queueUserNotification(user, l, added, deleted, wg)
 	}
 }
 
@@ -529,15 +536,18 @@ func (l *labelIdentitySelector) matches(identity scIdentity) bool {
 
 // UpdateFQDNSelector updates the mapping of fqdnKey (the FQDNSelector from a
 // policy rule as a string) to to the provided list of identities. If the contents
-// of the cachedSelections differ from those in the identities slice, all
-// users are notified.
-func (sc *SelectorCache) UpdateFQDNSelector(fqdnSelec api.FQDNSelector, identities []identity.NumericIdentity) {
+// of the cachedSelections differ from those in the identities slice, all users
+// are notified asynchronously. Caller should Wait() on the returned
+// sync.WaitGroup before triggering any policy updates. Policy updates may need
+// Endpoint locks, so this Wait() can deadlock if the caller is holding any
+// endpoint locks.
+func (sc *SelectorCache) UpdateFQDNSelector(fqdnSelec api.FQDNSelector, identities []identity.NumericIdentity, wg *sync.WaitGroup) {
 	sc.mutex.Lock()
-	sc.updateFQDNSelector(fqdnSelec, identities)
+	sc.updateFQDNSelector(fqdnSelec, identities, wg)
 	sc.mutex.Unlock()
 }
 
-func (sc *SelectorCache) updateFQDNSelector(fqdnSelec api.FQDNSelector, identities []identity.NumericIdentity) {
+func (sc *SelectorCache) updateFQDNSelector(fqdnSelec api.FQDNSelector, identities []identity.NumericIdentity, wg *sync.WaitGroup) {
 	fqdnKey := fqdnSelec.String()
 
 	var fqdnSel *fqdnSelector
@@ -621,7 +631,7 @@ func (sc *SelectorCache) updateFQDNSelector(fqdnSelec api.FQDNSelector, identiti
 	// getting the CIDR identities which correspond to this FQDNSelector. This
 	// is the primary difference here between FQDNSelector and IdentitySelector.
 	fqdnSel.updateSelections()
-	fqdnSel.notifyUsers(sc, added, deleted) // disjoint sets, see the comment above
+	fqdnSel.notifyUsers(sc, added, deleted, wg) // disjoint sets, see the comment above
 }
 
 // AddFQDNSelector adds the given api.FQDNSelector in to the selector cache. If
@@ -824,7 +834,11 @@ func (sc *SelectorCache) ChangeUser(selector CachedSelector, from, to CachedSele
 //
 // The caller is responsible for making sure the same identity is not
 // present in both 'added' and 'deleted'.
-func (sc *SelectorCache) UpdateIdentities(added, deleted cache.IdentityCache) {
+//
+// Caller should Wait() on the returned sync.WaitGroup before triggering any
+// policy updates. Policy updates may need Endpoint locks, so this Wait() can
+// deadlock if the caller is holding any endpoint locks.
+func (sc *SelectorCache) UpdateIdentities(added, deleted cache.IdentityCache, wg *sync.WaitGroup) {
 	sc.mutex.Lock()
 	defer sc.mutex.Unlock()
 
@@ -880,7 +894,7 @@ func (sc *SelectorCache) UpdateIdentities(added, deleted cache.IdentityCache) {
 				}
 				if len(dels)+len(adds) > 0 {
 					idSel.updateSelections()
-					idSel.notifyUsers(sc, adds, dels)
+					idSel.notifyUsers(sc, adds, dels, wg)
 				}
 			case *fqdnSelector:
 				// This is a no-op right now. We don't encode in the identities
@@ -892,12 +906,12 @@ func (sc *SelectorCache) UpdateIdentities(added, deleted cache.IdentityCache) {
 
 // RemoveIdentitiesFQDNSelectors removes all identities from being mapped to the
 // set of FQDNSelectors.
-func (sc *SelectorCache) RemoveIdentitiesFQDNSelectors(fqdnSels []api.FQDNSelector) {
+func (sc *SelectorCache) RemoveIdentitiesFQDNSelectors(fqdnSels []api.FQDNSelector, wg *sync.WaitGroup) {
 	sc.mutex.Lock()
 	noIdentities := []identity.NumericIdentity{}
 
 	for i := range fqdnSels {
-		sc.updateFQDNSelector(fqdnSels[i], noIdentities)
+		sc.updateFQDNSelector(fqdnSels[i], noIdentities, wg)
 	}
 	sc.mutex.Unlock()
 }

--- a/pkg/policy/selectorcache_test.go
+++ b/pkg/policy/selectorcache_test.go
@@ -17,11 +17,14 @@
 package policy
 
 import (
+	"sync"
+
 	"github.com/cilium/cilium/pkg/checker"
 	"github.com/cilium/cilium/pkg/identity"
 	"github.com/cilium/cilium/pkg/identity/cache"
 	k8sConst "github.com/cilium/cilium/pkg/k8s/apis/cilium.io"
 	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/lock"
 	"github.com/cilium/cilium/pkg/policy/api"
 	"github.com/cilium/cilium/pkg/testutils"
 
@@ -38,22 +41,33 @@ func (d *DummySelectorCacheUser) IdentitySelectionUpdated(selector CachedSelecto
 }
 
 type cachedSelectionUser struct {
-	c             *C
-	sc            *SelectorCache
-	name          string
+	c    *C
+	sc   *SelectorCache
+	name string
+
+	updateMutex   lock.Mutex
+	updateCond    *sync.Cond
 	selections    map[CachedSelector][]identity.NumericIdentity
 	notifications int
 	adds          int
 	deletes       int
 }
 
+func (sc *SelectorCache) haveUserNotifications() bool {
+	sc.userMutex.Lock()
+	defer sc.userMutex.Unlock()
+	return len(sc.userNotes) > 0
+}
+
 func newUser(c *C, name string, sc *SelectorCache) *cachedSelectionUser {
-	return &cachedSelectionUser{
+	csu := &cachedSelectionUser{
 		c:          c,
 		sc:         sc,
 		name:       name,
 		selections: make(map[CachedSelector][]identity.NumericIdentity),
 	}
+	csu.updateCond = sync.NewCond(&csu.updateMutex)
+	return csu
 }
 
 func haveNid(nid identity.NumericIdentity, selections []identity.NumericIdentity) bool {
@@ -66,7 +80,9 @@ func haveNid(nid identity.NumericIdentity, selections []identity.NumericIdentity
 }
 
 func (csu *cachedSelectionUser) AddIdentitySelector(sel api.EndpointSelector) CachedSelector {
-	notifications := csu.notifications
+	csu.updateMutex.Lock()
+	defer csu.updateMutex.Unlock()
+
 	cached, added := csu.sc.AddIdentitySelector(csu, sel)
 	csu.c.Assert(cached, Not(Equals), nil)
 
@@ -76,13 +92,15 @@ func (csu *cachedSelectionUser) AddIdentitySelector(sel api.EndpointSelector) Ca
 	csu.selections[cached] = cached.GetSelections()
 
 	// Pre-existing selections are not notified as updates
-	csu.c.Assert(csu.notifications, Equals, notifications)
+	csu.c.Assert(csu.sc.haveUserNotifications(), Equals, false)
 
 	return cached
 }
 
 func (csu *cachedSelectionUser) AddFQDNSelector(sel api.FQDNSelector) CachedSelector {
-	notifications := csu.notifications
+	csu.updateMutex.Lock()
+	defer csu.updateMutex.Unlock()
+
 	cached, added := csu.sc.AddFQDNSelector(csu, sel)
 	csu.c.Assert(cached, Not(Equals), nil)
 
@@ -92,21 +110,41 @@ func (csu *cachedSelectionUser) AddFQDNSelector(sel api.FQDNSelector) CachedSele
 	csu.selections[cached] = cached.GetSelections()
 
 	// Pre-existing selections are not notified as updates
-	csu.c.Assert(csu.notifications, Equals, notifications)
+	csu.c.Assert(csu.sc.haveUserNotifications(), Equals, false)
 
 	return cached
 }
 
 func (csu *cachedSelectionUser) RemoveSelector(sel CachedSelector) {
-	notifications := csu.notifications
+	csu.updateMutex.Lock()
+	defer csu.updateMutex.Unlock()
+
 	csu.sc.RemoveSelector(sel, csu)
 	delete(csu.selections, sel)
 
 	// No notifications for a removed selector
-	csu.c.Assert(csu.notifications, Equals, notifications)
+	csu.c.Assert(csu.sc.haveUserNotifications(), Equals, false)
+}
+
+func (csu *cachedSelectionUser) Reset() {
+	csu.updateMutex.Lock()
+	defer csu.updateMutex.Unlock()
+	csu.notifications = 0
+}
+
+func (csu *cachedSelectionUser) WaitForUpdate() (adds, deletes int) {
+	csu.updateMutex.Lock()
+	defer csu.updateMutex.Unlock()
+	for csu.notifications == 0 {
+		csu.updateCond.Wait()
+	}
+	return csu.adds, csu.deletes
 }
 
 func (csu *cachedSelectionUser) IdentitySelectionUpdated(selector CachedSelector, added, deleted []identity.NumericIdentity) {
+	csu.updateMutex.Lock()
+	defer csu.updateMutex.Unlock()
+
 	csu.notifications++
 	csu.adds += len(added)
 	csu.deletes += len(deleted)
@@ -123,6 +161,8 @@ func (csu *cachedSelectionUser) IdentitySelectionUpdated(selector CachedSelector
 
 	// update selections
 	csu.selections[selector] = selections
+
+	csu.updateCond.Signal()
 }
 
 // Mock CachedSelector for unit testing.
@@ -323,12 +363,14 @@ func (ds *SelectorCacheTestSuite) TestIdentityUpdates(c *C) {
 	c.Assert(len(selections2), Equals, 1)
 	c.Assert(selections2[0], Equals, identity.NumericIdentity(2345))
 
+	user1.Reset()
 	// Add some identities to the identity cache
 	sc.UpdateIdentities(cache.IdentityCache{
 		12345: labels.Labels{"app": labels.NewLabel("app", "test", labels.LabelSourceK8s)}.LabelArray(),
 	}, nil)
-	c.Assert(user1.adds, Equals, 1)
-	c.Assert(user1.deletes, Equals, 0)
+	adds, deletes := user1.WaitForUpdate()
+	c.Assert(adds, Equals, 1)
+	c.Assert(deletes, Equals, 0)
 
 	// Current selections contain the numeric identities of existing identities that match
 	selections = cached.GetSelections()
@@ -336,12 +378,14 @@ func (ds *SelectorCacheTestSuite) TestIdentityUpdates(c *C) {
 	c.Assert(selections[0], Equals, identity.NumericIdentity(1234))
 	c.Assert(selections[1], Equals, identity.NumericIdentity(12345))
 
+	user1.Reset()
 	// Remove some identities from the identity cache
 	sc.UpdateIdentities(nil, cache.IdentityCache{
 		12345: labels.Labels{"app": labels.NewLabel("app", "test", labels.LabelSourceK8s)}.LabelArray(),
 	})
-	c.Assert(user1.adds, Equals, 1)
-	c.Assert(user1.deletes, Equals, 1)
+	adds, deletes = user1.WaitForUpdate()
+	c.Assert(adds, Equals, 1)
+	c.Assert(deletes, Equals, 1)
 
 	// Current selections contain the numeric identities of existing identities that match
 	selections = cached.GetSelections()
@@ -392,19 +436,26 @@ func (ds *SelectorCacheTestSuite) TestFQDNSelectorUpdates(c *C) {
 	}
 
 	// Add some identities to the identity cache
+	user1.Reset()
 	ciliumIdentities = append(ciliumIdentities, identity.NumericIdentity(123456))
 	sc.UpdateFQDNSelector(ciliumSel, ciliumIdentities)
-	c.Assert(user1.adds, Equals, 1)
-	c.Assert(user1.deletes, Equals, 0)
+	adds, deletes := user1.WaitForUpdate()
+	c.Assert(adds, Equals, 1)
+	c.Assert(deletes, Equals, 0)
 
+	user1.Reset()
 	ciliumIdentities = ciliumIdentities[:1]
 	sc.UpdateFQDNSelector(ciliumSel, ciliumIdentities)
-	c.Assert(user1.adds, Equals, 1)
-	c.Assert(user1.deletes, Equals, 3)
+	adds, deletes = user1.WaitForUpdate()
+	c.Assert(adds, Equals, 1)
+	c.Assert(deletes, Equals, 3)
 
+	user1.Reset()
 	ciliumIdentities = []identity.NumericIdentity{}
 	sc.UpdateFQDNSelector(ciliumSel, ciliumIdentities)
-	c.Assert(user1.deletes, Equals, 4)
+	adds, deletes = user1.WaitForUpdate()
+	c.Assert(adds, Equals, 1)
+	c.Assert(deletes, Equals, 4)
 
 	user1.RemoveSelector(cached)
 	user1.RemoveSelector(cached2)
@@ -484,16 +535,20 @@ func (ds *SelectorCacheTestSuite) TestIdentityUpdatesMultipleUsers(c *C) {
 	cached2 := user2.AddIdentitySelector(testSelector)
 	c.Assert(cached2, Equals, cached)
 
+	user1.Reset()
+	user2.Reset()
 	// Add some identities to the identity cache
 	sc.UpdateIdentities(cache.IdentityCache{
 		123: labels.Labels{"app": labels.NewLabel("app", "test", labels.LabelSourceK8s)}.LabelArray(),
 		234: labels.Labels{"app": labels.NewLabel("app", "test2", labels.LabelSourceK8s)}.LabelArray(),
 		345: labels.Labels{"app": labels.NewLabel("app", "test", labels.LabelSourceK8s)}.LabelArray(),
 	}, nil)
-	c.Assert(user1.adds, Equals, 2)
-	c.Assert(user1.deletes, Equals, 0)
-	c.Assert(user2.adds, Equals, 2)
-	c.Assert(user2.deletes, Equals, 0)
+	adds, deletes := user1.WaitForUpdate()
+	c.Assert(adds, Equals, 2)
+	c.Assert(deletes, Equals, 0)
+	adds, deletes = user2.WaitForUpdate()
+	c.Assert(adds, Equals, 2)
+	c.Assert(deletes, Equals, 0)
 
 	// Current selections contain the numeric identities of existing identities that match
 	selections := cached.GetSelections()
@@ -504,15 +559,19 @@ func (ds *SelectorCacheTestSuite) TestIdentityUpdatesMultipleUsers(c *C) {
 
 	c.Assert(cached.GetSelections(), checker.DeepEquals, cached2.GetSelections())
 
+	user1.Reset()
+	user2.Reset()
 	// Remove some identities from the identity cache
 	sc.UpdateIdentities(nil, cache.IdentityCache{
 		123: labels.Labels{"app": labels.NewLabel("app", "test", labels.LabelSourceK8s)}.LabelArray(),
 		234: labels.Labels{"app": labels.NewLabel("app", "test2", labels.LabelSourceK8s)}.LabelArray(),
 	})
-	c.Assert(user1.adds, Equals, 2)
-	c.Assert(user1.deletes, Equals, 1)
-	c.Assert(user2.adds, Equals, 2)
-	c.Assert(user2.deletes, Equals, 1)
+	adds, deletes = user1.WaitForUpdate()
+	c.Assert(adds, Equals, 2)
+	c.Assert(deletes, Equals, 1)
+	adds, deletes = user2.WaitForUpdate()
+	c.Assert(adds, Equals, 2)
+	c.Assert(deletes, Equals, 1)
 
 	// Current selections contain the numeric identities of existing identities that match
 	selections = cached.GetSelections()

--- a/pkg/policy/selectorcache_test.go
+++ b/pkg/policy/selectorcache_test.go
@@ -252,11 +252,13 @@ func (ds *SelectorCacheTestSuite) TestAddRemoveSelector(c *C) {
 	sc := testNewSelectorCache(cache.IdentityCache{})
 
 	// Add some identities to the identity cache
+	wg := &sync.WaitGroup{}
 	sc.UpdateIdentities(cache.IdentityCache{
 		1234: labels.Labels{"app": labels.NewLabel("app", "test", labels.LabelSourceK8s),
 			k8sConst.PodNamespaceLabel: labels.NewLabel(k8sConst.PodNamespaceLabel, "default", labels.LabelSourceK8s)}.LabelArray(),
 		2345: labels.Labels{"app": labels.NewLabel("app", "test2", labels.LabelSourceK8s)}.LabelArray(),
-	}, nil)
+	}, nil, wg)
+	wg.Wait()
 
 	testSelector := api.NewESFromLabels(labels.NewLabel("app", "test", labels.LabelSourceK8s),
 		labels.NewLabel(k8sConst.PodNamespaceLabel, "default", labels.LabelSourceK8s))
@@ -302,10 +304,12 @@ func (ds *SelectorCacheTestSuite) TestMultipleIdentitySelectors(c *C) {
 	sc := testNewSelectorCache(cache.IdentityCache{})
 
 	// Add some identities to the identity cache
+	wg := &sync.WaitGroup{}
 	sc.UpdateIdentities(cache.IdentityCache{
 		1234: labels.Labels{"app": labels.NewLabel("app", "test", labels.LabelSourceK8s)}.LabelArray(),
 		2345: labels.Labels{"app": labels.NewLabel("app", "test2", labels.LabelSourceK8s)}.LabelArray(),
-	}, nil)
+	}, nil, wg)
+	wg.Wait()
 
 	testSelector := api.NewESFromLabels(labels.NewLabel("app", "test", labels.LabelSourceK8s))
 	test2Selector := api.NewESFromLabels(labels.NewLabel("app", "test2", labels.LabelSourceK8s))
@@ -338,10 +342,12 @@ func (ds *SelectorCacheTestSuite) TestIdentityUpdates(c *C) {
 	sc := testNewSelectorCache(cache.IdentityCache{})
 
 	// Add some identities to the identity cache
+	wg := &sync.WaitGroup{}
 	sc.UpdateIdentities(cache.IdentityCache{
 		1234: labels.Labels{"app": labels.NewLabel("app", "test", labels.LabelSourceK8s)}.LabelArray(),
 		2345: labels.Labels{"app": labels.NewLabel("app", "test2", labels.LabelSourceK8s)}.LabelArray(),
-	}, nil)
+	}, nil, wg)
+	wg.Wait()
 
 	testSelector := api.NewESFromLabels(labels.NewLabel("app", "test", labels.LabelSourceK8s))
 	test2Selector := api.NewESFromLabels(labels.NewLabel("app", "test2", labels.LabelSourceK8s))
@@ -365,9 +371,12 @@ func (ds *SelectorCacheTestSuite) TestIdentityUpdates(c *C) {
 
 	user1.Reset()
 	// Add some identities to the identity cache
+	wg = &sync.WaitGroup{}
 	sc.UpdateIdentities(cache.IdentityCache{
 		12345: labels.Labels{"app": labels.NewLabel("app", "test", labels.LabelSourceK8s)}.LabelArray(),
-	}, nil)
+	}, nil, wg)
+	wg.Wait()
+
 	adds, deletes := user1.WaitForUpdate()
 	c.Assert(adds, Equals, 1)
 	c.Assert(deletes, Equals, 0)
@@ -380,9 +389,12 @@ func (ds *SelectorCacheTestSuite) TestIdentityUpdates(c *C) {
 
 	user1.Reset()
 	// Remove some identities from the identity cache
+	wg = &sync.WaitGroup{}
 	sc.UpdateIdentities(nil, cache.IdentityCache{
 		12345: labels.Labels{"app": labels.NewLabel("app", "test", labels.LabelSourceK8s)}.LabelArray(),
-	})
+	}, wg)
+	wg.Wait()
+
 	adds, deletes = user1.WaitForUpdate()
 	c.Assert(adds, Equals, 1)
 	c.Assert(deletes, Equals, 1)
@@ -409,8 +421,10 @@ func (ds *SelectorCacheTestSuite) TestFQDNSelectorUpdates(c *C) {
 	googleIdentities := []identity.NumericIdentity{321, 456, 987}
 	ciliumIdentities := []identity.NumericIdentity{123, 456, 789}
 
-	sc.UpdateFQDNSelector(ciliumSel, ciliumIdentities)
-	sc.UpdateFQDNSelector(googleSel, googleIdentities)
+	wg := &sync.WaitGroup{}
+	sc.UpdateFQDNSelector(ciliumSel, ciliumIdentities, wg)
+	sc.UpdateFQDNSelector(googleSel, googleIdentities, wg)
+	wg.Wait()
 
 	_, exists := sc.selectors[ciliumSel.String()]
 	c.Assert(exists, Equals, true)
@@ -438,21 +452,30 @@ func (ds *SelectorCacheTestSuite) TestFQDNSelectorUpdates(c *C) {
 	// Add some identities to the identity cache
 	user1.Reset()
 	ciliumIdentities = append(ciliumIdentities, identity.NumericIdentity(123456))
-	sc.UpdateFQDNSelector(ciliumSel, ciliumIdentities)
+	wg = &sync.WaitGroup{}
+	sc.UpdateFQDNSelector(ciliumSel, ciliumIdentities, wg)
+	wg.Wait()
+
 	adds, deletes := user1.WaitForUpdate()
 	c.Assert(adds, Equals, 1)
 	c.Assert(deletes, Equals, 0)
 
 	user1.Reset()
 	ciliumIdentities = ciliumIdentities[:1]
-	sc.UpdateFQDNSelector(ciliumSel, ciliumIdentities)
+	wg = &sync.WaitGroup{}
+	sc.UpdateFQDNSelector(ciliumSel, ciliumIdentities, wg)
+	wg.Wait()
+
 	adds, deletes = user1.WaitForUpdate()
 	c.Assert(adds, Equals, 1)
 	c.Assert(deletes, Equals, 3)
 
 	user1.Reset()
 	ciliumIdentities = []identity.NumericIdentity{}
-	sc.UpdateFQDNSelector(ciliumSel, ciliumIdentities)
+	wg = &sync.WaitGroup{}
+	sc.UpdateFQDNSelector(ciliumSel, ciliumIdentities, wg)
+	wg.Wait()
+
 	adds, deletes = user1.WaitForUpdate()
 	c.Assert(adds, Equals, 1)
 	c.Assert(deletes, Equals, 4)
@@ -478,8 +501,10 @@ func (ds *SelectorCacheTestSuite) TestRemoveIdentitiesFQDNSelectors(c *C) {
 	googleIdentities := []identity.NumericIdentity{321, 456, 987}
 	ciliumIdentities := []identity.NumericIdentity{123, 456, 789}
 
-	sc.UpdateFQDNSelector(ciliumSel, ciliumIdentities)
-	sc.UpdateFQDNSelector(googleSel, googleIdentities)
+	wg := &sync.WaitGroup{}
+	sc.UpdateFQDNSelector(ciliumSel, ciliumIdentities, wg)
+	sc.UpdateFQDNSelector(googleSel, googleIdentities, wg)
+	wg.Wait()
 
 	_, exists := sc.selectors[ciliumSel.String()]
 	c.Assert(exists, Equals, true)
@@ -504,10 +529,12 @@ func (ds *SelectorCacheTestSuite) TestRemoveIdentitiesFQDNSelectors(c *C) {
 		c.Assert(selection, Equals, googleIdentities[i])
 	}
 
+	wg = &sync.WaitGroup{}
 	sc.RemoveIdentitiesFQDNSelectors([]api.FQDNSelector{
 		googleSel,
 		ciliumSel,
-	})
+	}, wg)
+	wg.Wait()
 
 	selections = cached.GetSelections()
 	c.Assert(len(selections), Equals, 0)
@@ -520,10 +547,12 @@ func (ds *SelectorCacheTestSuite) TestIdentityUpdatesMultipleUsers(c *C) {
 	sc := testNewSelectorCache(cache.IdentityCache{})
 
 	// Add some identities to the identity cache
+	wg := &sync.WaitGroup{}
 	sc.UpdateIdentities(cache.IdentityCache{
 		1234: labels.Labels{"app": labels.NewLabel("app", "test", labels.LabelSourceK8s)}.LabelArray(),
 		2345: labels.Labels{"app": labels.NewLabel("app", "test2", labels.LabelSourceK8s)}.LabelArray(),
-	}, nil)
+	}, nil, wg)
+	wg.Wait()
 
 	testSelector := api.NewESFromLabels(labels.NewLabel("app", "test", labels.LabelSourceK8s))
 
@@ -538,11 +567,14 @@ func (ds *SelectorCacheTestSuite) TestIdentityUpdatesMultipleUsers(c *C) {
 	user1.Reset()
 	user2.Reset()
 	// Add some identities to the identity cache
+	wg = &sync.WaitGroup{}
 	sc.UpdateIdentities(cache.IdentityCache{
 		123: labels.Labels{"app": labels.NewLabel("app", "test", labels.LabelSourceK8s)}.LabelArray(),
 		234: labels.Labels{"app": labels.NewLabel("app", "test2", labels.LabelSourceK8s)}.LabelArray(),
 		345: labels.Labels{"app": labels.NewLabel("app", "test", labels.LabelSourceK8s)}.LabelArray(),
-	}, nil)
+	}, nil, wg)
+	wg.Wait()
+
 	adds, deletes := user1.WaitForUpdate()
 	c.Assert(adds, Equals, 2)
 	c.Assert(deletes, Equals, 0)
@@ -562,10 +594,13 @@ func (ds *SelectorCacheTestSuite) TestIdentityUpdatesMultipleUsers(c *C) {
 	user1.Reset()
 	user2.Reset()
 	// Remove some identities from the identity cache
+	wg = &sync.WaitGroup{}
 	sc.UpdateIdentities(nil, cache.IdentityCache{
 		123: labels.Labels{"app": labels.NewLabel("app", "test", labels.LabelSourceK8s)}.LabelArray(),
 		234: labels.Labels{"app": labels.NewLabel("app", "test2", labels.LabelSourceK8s)}.LabelArray(),
-	})
+	}, wg)
+	wg.Wait()
+
 	adds, deletes = user1.WaitForUpdate()
 	c.Assert(adds, Equals, 2)
 	c.Assert(deletes, Equals, 1)
@@ -625,10 +660,12 @@ func (ds *SelectorCacheTestSuite) TestIdentityNotifier(c *C) {
 	selections2 := cached2.GetSelections()
 	c.Assert(len(selections2), Equals, 0)
 
+	wg := &sync.WaitGroup{}
 	sc.RemoveIdentitiesFQDNSelectors([]api.FQDNSelector{
 		googleSel,
 		ciliumSel,
-	})
+	}, wg)
+	wg.Wait()
 
 	selections = cached.GetSelections()
 	c.Assert(len(selections), Equals, 0)

